src/shogun/clustering/KMeans.cpp

```diff
@@ -4,446 +4,189 @@
 #include <shogun/distance/EuclideanDistance.h>
-#include <shogun/labels/Labels.h>
 #include <shogun/features/DenseFeatures.h>
 #include <shogun/mathematics/Math.h>
-#include <shogun/base/Parallel.h>
+#include <shogun/io/SGIO.h>
 #include <shogun/mathematics/eigen3.h>
-
-#ifdef HAVE_PTHREAD
-#include <pthread.h>
-#endif
-
-#ifdef HAVE_LINALG_LIB
-#include <shogun/mathematics/linalg/linalg.h>
-#endif
-
-using namespace shogun;
+#include <iostream>
 using namespace Eigen;
+using namespace shogun;
 
-CKMeans::CKMeans()
-: CDistanceMachine()
+
+namespace shogun
 {
-	init();
-}
 
-CKMeans::CKMeans(int32_t k_, CDistance* d, EKMeansMethod f)
-:CDistanceMachine()
+CKMeans::CKMeans():CKMeansBase()
 {
-	init();
-	k=k_;
-	set_distance(d);
-	train_method=f;
 }
 
-CKMeans::CKMeans(int32_t k_, CDistance* d, bool use_kmpp, EKMeansMethod f)
-: CDistanceMachine()
+CKMeans::CKMeans(int32_t k_i, CDistance* d_i, bool use_kmpp_i):CKMeansBase(k_i, d_i, use_kmpp_i)
 {
-	init();
-	k=k_;
-	set_distance(d);
-	use_kmeanspp=use_kmpp;
-	train_method=f;
 }
 
-CKMeans::CKMeans(int32_t k_i, CDistance* d_i, SGMatrix<float64_t> centers_i, EKMeansMethod f)
-: CDistanceMachine()
+CKMeans::CKMeans(int32_t k_i, CDistance* d_i, SGMatrix<float64_t> centers_i):CKMeansBase(k_i, d_i, centers_i)
 {
-	init();
-	k = k_i;
-	set_distance(d_i);
-	set_initial_centers(centers_i);
-	train_method=f;
 }
 
 CKMeans::~CKMeans()
 {
 }
 
-void CKMeans::set_initial_centers(SGMatrix<float64_t> centers)
+void CKMeans::Lloyd_KMeans(SGMatrix<float64_t> centers, int32_t num_centers)
 {
- 	CDenseFeatures<float64_t>* lhs=((CDenseFeatures<float64_t>*) distance->get_lhs());
-	dimensions=lhs->get_num_features();
-	REQUIRE(centers.num_cols == k,
-			"Expected %d initial cluster centers, got %d.\n", k, centers.num_cols);
-	REQUIRE(centers.num_rows == dimensions,
-			"Expected %d dimensionional cluster centers, got %d.\n", dimensions, centers.num_rows);
-	mus_initial = centers;
-	SG_UNREF(lhs);
-}
-
-void CKMeans::set_random_centers()
-{
-	mus.zero();
 	CDenseFeatures<float64_t>* lhs=
 		CDenseFeatures<float64_t>::obtain_from_generic(distance->get_lhs());
-	int32_t lhs_size=lhs->get_num_vectors();
 	
-	SGVector<int32_t> temp=SGVector<int32_t>(lhs_size);
-	SGVector<int32_t>::range_fill_vector(temp, lhs_size, 0);
-	CMath::permute(temp);
+	int32_t lhs_size=lhs->get_num_vectors();
+	int32_t dim=lhs->get_num_features();
 
-	for (int32_t i=0; i<k; i++)
-	{
-		const int32_t cluster_center_i=temp[i];
-		SGVector<float64_t> vec=lhs->get_feature_vector(cluster_center_i);
+	CDenseFeatures<float64_t>* rhs_mus=new CDenseFeatures<float64_t>(0);
+	CFeatures* rhs_cache=distance->replace_rhs(rhs_mus);
 
-		for (int32_t j=0; j<dimensions; j++)
-			mus(j,i)=vec[j];
+	SGVector<int32_t> cluster_assignments=SGVector<int32_t>(lhs_size);
+	cluster_assignments.zero();
 
-		lhs->free_feature_vector(vec, cluster_center_i);
-	}
+	/* Weights : Number of points in each cluster */
+	SGVector<int32_t> weights_set(num_centers);
```

Comment:
is `2^31 = 2147483648` i.e. 2.1 billion points enough? :) i mean in case of a big data set it could be more than that or?
i'm just suggesting that maybe using `uint32_t` or rather `int64_t` or `uint64_t` would be better choice, or?